File: chap2/ConjugateGradientMethods.f90
Program ConjugateGradientMethods
  Implicit none
  Integer :: i
  Integer, parameter :: n = 3
  Real(kind=8) :: A(n,n) = [ 4.d0, -2.d0, 2.d0, -2.d0, 2.d0, -4.d0, 2.d0, -4.d0, 11.d0 ]
  Real(kind=8) :: b(n,1) = [ 6.d0,-10.d0,27.d0 ]
  Real(kind=8) :: alpha = 0.d0, beta = 0.d0
  Real(kind=8) :: x(n,1) = 0.d0, d(n,1) = 0.d0, r(n,1) = 0.d0, rtmp(n,1) = 0.d0, tmp(1,1) = 0.d0
  
  d = b - matmul( A,x )
  r = d

  Do i = 0, n - 1
    If ( all(r<0.0) ) then
      exit
    End if
    rtmp = r
    tmp = matmul( transpose(r),r ) / matmul( matmul( transpose(d),A ), d )
    alpha = tmp(1,1)
    x = x + alpha * d
    r = r - alpha * matmul( A,d )
    tmp = matmul( transpose(r),r ) / matmul( transpose(rtmp),rtmp )
    beta = tmp(1,1)
    d = r + beta * d
  End do
  
  Do i = 1, n
    Write ( *,'(f12.5)' ) x(i,1)
  End do
End program ConjugateGradientMethods

File: chap2/ConjugateGradientMethods.py
# -*- coding: utf-8 -*-
"""
Created on Fri May 18 10:57:51 2018

@author: luk
"""

import numpy as np
n = 3
a = np.array( [[4., -2., 2. ],[-2., 2., -4.],[2., -4., 11.]], dtype = np.float64 )
b = np.array( [6., -10., 27.], dtype = np.float64 )
alpha = 0.
beta = 0.
x = np.zeros( (n,), dtype = np.float64 )
d = np.zeros( (n,), dtype = np.float64 )
r = np.zeros_like( x, dtype = np.float64 )
rtmp = np.zeros_like( r, dtype = np.float64 )

d = b - np.matmul( a,x )
r = d

for i in range(n):
    if ( all( r<0.0 ) ):
        quit()
    rtmp = r
    tmp = np.matmul( np.transpose(r),r ) / np.matmul( np.matmul( np.transpose(d), a ), d )
    alpha = tmp
    x = x + alpha * d
    r = r - alpha * np.matmul( a,d )
    tmp = np.matmul( np.transpose(r),r ) / np.matmul( np.transpose(rtmp), rtmp ) 
    beta = tmp
    d = r + beta * d

print( "the x of ConjugateGradientMethods is :" )  
for i in range(n):
    print( x[i] )
    
print( "the x of python is :" )  
a = np.linalg.inv(a)  # 求矩阵a的逆矩阵
x = np.matmul(a,b)
for i in range(len(x)):
      print( x[i] )
     
print( "please input Ente and stop!" )
input()

    